verification/curator-service/api/src/geocoding/adminsFetcher.ts

```diff
@@ -0,0 +1,95 @@
+import { GeocodeResult, Resolution } from './geocoder';
+
+import axios from 'axios';
+
+/**
+ * Mapbox administrative area fetcher using the boundaries API.
+ * https://www.mapbox.com/boundaries/.
+ *
+ * Some geocoding results do not contain the full administrative area details,
+ * such details can be subsequently fetched using the boundaries API by doing a
+ * tilequery fetch of the center of the geocode result with the administrative
+ * boundaries layers.
+ * Cf. https://docs.mapbox.com/help/glossary/tilequery-api/
+ *
+ * TODO: Explain the custom local fetches of names once implemented.
+ */
+export default class MapboxAdminsFetcher {
+    constructor(private readonly accessToken: string) {}
+
+    // Fill in missing admin levels for the given GeocodeResult.
+    async fillAdmins(geocode: GeocodeResult): Promise<void> {
+        // Get missing admins from geocode result.
+        const missingAdmins = new Set<Resolution>();
+        if (geocode.administrativeAreaLevel1 === '') {
+            missingAdmins.add(Resolution.Admin1);
+        }
+        if (geocode.administrativeAreaLevel2 === '') {
+            missingAdmins.add(Resolution.Admin2);
+        }
+        if (geocode.administrativeAreaLevel3 === '') {
+            missingAdmins.add(Resolution.Admin3);
+        }
+        if (missingAdmins.size === 0) {
+            return;
+        }
+        // Fetch all missing admins in one query.
+        const url = `https://api.mapbox.com/v4/${getLayersParam(
+            missingAdmins,
+        )}/tilequery/${geocode.geometry.longitude},${
+            geocode.geometry.latitude
+        }.json?access_token=${this.accessToken}`;
+        try {
+            const resp = await axios.get(url);
+            const features = (resp.data as BoundariesResponse).features;
+            features.forEach((feature) => {
+                switch (feature.properties.tilequery.layer) {
+                    case 'boundaries_admin_1':
+                    // TODO: Get name from prop.id and put in geocode.administrativeAreaLevel1;
+                    case 'boundaries_admin_2':
+                    // TODO: Get name from prop.id and put in geocode.administrativeAreaLevel2;
+                    case 'boundaries_admin_3':
+                    // TODO: Get name from prop.id and put in geocode.administrativeAreaLevel3;
+                }
+            });
+        } catch (e) {
+            // Fail gracefully, not being able to fetch all admins isn't a huge deal.
```

Comment:
> isn't a huge deal.

I'm not totally sure. Do we want case entry to fail (e.g. for bulk upload) if "Adams, Ohio" doesn't find an admin2?

Reply:
For manual entry it's definitely optional (as one can still enter it in the UI if not found), if we want to require it we should also make it mandatory in the schema but honestly I doubt we can find ALL admin levels EVERYWHERE in the world that seems like a pretty strong statement... I'd feel better getting 99% of them and not making it mandatory, perhaps there can be a manual process to fix those missing afterwards.

For bulk upload we might want to fail early and have the users resubmit the csv, I am not sure, either way that sounds like a request param which we can add in a later PR to make this mandatory if we decide to, sg?

Reply:
Yep, SG!